### PR TITLE
client-gen: create private registry in fake clientset

### PR DESCRIPTION
**Pull request by sttts**
This cuts of the laster `k8s.io/kubernetes/pkg/api.{Registry+Scheme+Codecs}` dependency from the clientsets. This enables clientset generation for packages that must not have a dependency onto kubernetes itself.

@deads2k there is more than the namespace checking we discussed: the RESTMapper built from the registry. This introduces a private registry. I try get that out from the normal versioned client as much as possible. I would even like to remove this private registry some day, at best remove all registry code from the client. But that's for another day...

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh [hunk 1]
```diff
@@ -77,13 +77,3 @@ ${informergen} \
   --listers-package k8s.io/kube-aggregator/pkg/client/listers \
   --output-package k8s.io/kube-aggregator/pkg/client/informers
   "$@"
-
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: @deads2k isn't this a great pain relieve?

Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh]: > @deads2k isn't this a great pain relieve?

That feel good.

File: federation/client/clientset_generated/federation_clientset/fake/register.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	batch "k8s.io/kubernetes/pkg/apis/batch/install"
+	extensions "k8s.io/kubernetes/pkg/apis/extensions/install"
+	os "os"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = serializer.NewCodecFactory(scheme)
+var parameterCodec = runtime.NewParameterCodec(scheme)
+
+var registry = registered.NewOrDie(os.Getenv("KUBE_API_VERSIONS"))
```
Comment (deads2k) [inline federation/client/clientset_generated/federation_clientset/fake/register.go]: We should kill this env var next release.

Comment (sttts) [inline federation/client/clientset_generated/federation_clientset/fake/register.go]: Yes. Or at least make a client behave gracefully. Panic is not gracefully.

**Comment by smarterclayton** [comment]
Fakes can panic - no one should be using a fake in normal code?

**Comment by sttts** [comment]
@smarterclayton for fakes it's fine. But we also have code like the apiserver bootstrapping where the behaviour is totally undefined right now with the env var, informers crash etc.

**Comment by sttts** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
flaked on https://github.com/kubernetes/kubernetes/issues/41892

**Comment by sttts** [comment]
Rebased and updated the client.

**Comment by caesarxuchao** [comment]
> there is more than the namespace checking we discussed: the RESTMapper built from the registry.

To stop copying types to client-go, we'll need to get rid of the private registry from the versioned fake client, as that imports the internal types, which will not be moved to [k8s.io/api](https://github.com/kubernetes/kubernetes/pull/44784)

**Comment by caesarxuchao** [comment]
@sttts @deads2k i plan to remove the [namespace checking](https://github.com/kubernetes/kubernetes/blob/ae1c9a2b25ddc043b2b93d4ef11bacc02d4fd55f/staging/src/k8s.io/client-go/testing/fixture.go#L415-L418) in testing/fixture.go. The fake clientset is generated, we don't need to verify if the action generated by the fake clientset is correctly namespaced in each test that uses the fake clientset.

I also plan to remove the use of the [RESTMapper](https://github.com/kubernetes/kubernetes/blob/ae1c9a2b25ddc043b2b93d4ef11bacc02d4fd55f/staging/src/k8s.io/client-go/testing/fixture.go#L415-L418) in testing/fixture.go. The fixture.go should just use GroupVersionResource and shoulnd't use restmapper to translate to GVK. The only necessary use of GVK in fixture.go is [guessing the GVK of a list](https://github.com/kubernetes/kubernetes/blob/ae1c9a2b25ddc043b2b93d4ef11bacc02d4fd55f/staging/src/k8s.io/client-go/testing/fixture.go#L168-L172). We can just add a ListKind field to the `ListActionImpl` and ask the test writers to fill it.

**Comment by caesarxuchao** [comment]
Sent #45159. Still WIP.

**Pull request merged**